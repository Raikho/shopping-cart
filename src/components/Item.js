const Item = props => {
    const { handleChangeQuantity } = props;
    const { name, price, id, img, quantity } = props.item;

    const onSubtract = () => {handleChangeQuantity(id, -1);}
    const onAdd = () => {handleChangeQuantity(id, 1);}

    return (
        <div className="item">
            <div className="name">{name}</div>
            <img src={require(`../assets/${img}.jpg`)} alt={img} />
            <div className="price">{`$${price}.00`}</div>
            {quantity > 0 ?
                <div>
                    <button onClick={onSubtract}>-</button>
                    {quantity}
                    <button onClick={onAdd}>+</button>
                </div> :
                <button onClick={onAdd}>Buy</button>
            }
        </div>
    )
}

export default Item;