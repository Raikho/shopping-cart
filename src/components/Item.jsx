import React from "react";

export default function Item({ item, onChangeAmount }) {
    const { name, price, id, img, quantity } = item;

    let imgUrl = new URL(`../assets/${img}.jpg`, import.meta.url);
    const onAdd = () => {onChangeAmount(id, 1)};
    const onSubtract = () => {onChangeAmount(id, -1)};

    return (
        <div className="item">
            <img src={imgUrl} alt={img} />
            <div className="desc">
                <div className="name">{name}</div>
                <div className="price">{`$${price}.00`}</div>
                {quantity > 0 ?
                    <div className='quantity-container'>
                        <button className='adjust' onClick={onSubtract}>-</button>
                        {quantity}
                        <button className='adjust' onClick={onAdd}>+</button>
                    </div> :
                    <button className='add' onClick={onAdd}>Add</button>
                }
            </div>
        </div>
    );
}