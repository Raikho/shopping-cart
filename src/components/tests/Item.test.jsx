import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Item from '../Item';

const mockItem = {name: 'Betta', cat: 'fresh', price: 9, id: 123, img: 'betta', quantity: 0};

describe('Item component', () => {
    it('renders', () => {
        const { container } = render(<Item item={mockItem}/>)
        expect(container).toMatchSnapshot();
    });

    it('renders using given fields', () => {
        const item = mockItem;
        render(<Item item={item} />);
        expect(screen.getByText('Betta')).toBeInTheDocument();
        expect(screen.getByText('$9.00')).toBeInTheDocument();
    });

    it('renders correct buttons at 0 quantity', async () => {
        const user = userEvent.setup();
        const mock = vi.fn();
        render(<Item item={mockItem} onChangeAmount={mock}/>)

        const addButton = screen.getByRole("button", {name: "Add"});
        expect(screen.queryByText("+")).not.toBeInTheDocument();
        expect(screen.queryByText("-")).not.toBeInTheDocument();

        await user.click(addButton);
        expect(mock).toHaveBeenCalled();
    });

    it('renders correct buttons at 1+ quantity', async() => {
        const user = userEvent.setup();
        const mock = vi.fn();
        render(<Item item={{...mockItem, quantity: 9}}  onChangeAmount={mock}/>)

        expect(screen.queryByText("add")).not.toBeInTheDocument();
        expect(screen.queryByText("9")).toBeInTheDocument();
        const addButton = screen.getByRole("button", {name: "+"});
        const subtractButton = screen.getByRole("button", {name: "-"});

        await user.click(addButton);
        await user.click(subtractButton);
        expect(mock).toHaveBeenCalledTimes(2);
    });
});
