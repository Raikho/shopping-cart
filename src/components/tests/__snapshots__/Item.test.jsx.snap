// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item component > renders 1`] = `
<div>
  <div
    class="item"
  >
    <img
      alt="betta"
      src="file:///home/mateen/repos/shopping-cart/src/assets/betta.jpg"
    />
    <div
      class="desc"
    >
      <div
        class="name"
      >
        Betta
      </div>
      <div
        class="price"
      >
        $9.00
      </div>
      <button
        class="add"
      >
        Add
      </button>
    </div>
  </div>
</div>
`;
